barretenberg/cpp/src/barretenberg/relations/toy_avm/generic_lookup_relation.hpp

```diff
@@ -0,0 +1,469 @@
+            if (Settings::READ_TERM_TYPES[i] == READ_BASIC_TUPLE) {
+                current_degree = 1;
+            } else if (Settings::READ_TERM_TYPES[i] == READ_SCALED_TUPLE) {
+                current_degree = 2;
+            } else {
+                current_degree = Settings::READ_TERM_DEGREE;
+            }
+            accumulated_degree += current_degree;
+        }
+        return accumulated_degree;
+    }
+
+    /**
+     * @brief Compute the degree of of the product of write terms
+     *
+     * @details The degree of the inverse polynomial check subrelation is dependent on this value
+     *
+     * @return constexpr size_t
+     */
+    static constexpr size_t compute_write_term_product_degree()
+    {
+        size_t accumulated_degree = 0;
+        for (size_t i = 0; i < WRITE_TERMS; i++) {
+            size_t current_degree = 0;
+            if (Settings::WRITE_TERM_TYPES[i] == WRITE_BASIC_TUPLE) {
+                current_degree = 1;
+            } else {
+                current_degree = Settings::WRITE_TERM_DEGREE;
+            }
+            accumulated_degree += current_degree;
+        }
+        return accumulated_degree;
+    }
+
+    // Read term degree is dependent on what type of read term we use
+    static constexpr size_t READ_TERM_DEGREE = compute_maximum_read_term_degree();
+    static_assert(READ_TERM_DEGREE != 0);
+
+    // Write  term degree is dependent on what type of write term we use
+    static constexpr size_t WRITE_TERM_DEGREE = compute_maximum_write_term_degree();
+
+    static_assert(WRITE_TERM_DEGREE != 0);
+
+    // Compute the length of the inverse polynomial correctness sub-relation MAX(product of terms * inverse, inverse
+    // exists polynomial) + 1;
+    static constexpr size_t FIRST_SUBRELATION_LENGTH =
+        std::max((compute_read_term_product_degree() + compute_write_term_product_degree() + 1),
+                 Settings::INVERSE_EXISTS_POLYNOMIAL_DEGREE) +
+        1;
+
+    // Compute the length of the log-derived term subrelation MAX(read term * enable read, write term * write count *
+    // enable write)
+    static constexpr size_t SECOND_SUBRELATION_LENGTH = std::max(READ_TERM_DEGREE + 1, WRITE_TERM_DEGREE + 2);
+    // 1 + polynomial degree of this relation
+    static constexpr size_t LENGTH = std::max(FIRST_SUBRELATION_LENGTH, SECOND_SUBRELATION_LENGTH);
+
+    // The structure of polynomial tuple returned from Settings' functions get_const_entities and get_nonconst_entities
+    // is the following:
+    // 1) 1 Polynomial used to contain the inverse product from which we reconstruct individual inverses
+    // used in the sum
+    // 2) WRITE_TERMS number of polynomials representing how much each write term has been read
+    // 3) READ_TERMS number of polynomials enabling the addition of a particular read term in this row (should we lookup
+    // or not)
+    // 4) WRITE_TERMS number of polynomials enabling a particular write term in this row (should we add it to
```

Comment:
are 2 and 4 meant to be labelled the same?

Reply:
These are not labels, this is how many polynomials there are of each type (WRITE_TERMS) is how many table entries can be added per row